src/SAWScript/SolverCache/lmdb_opt_database.py

```diff
@@ -0,0 +1,202 @@
+       string as a prefix of their hex representation. If there is a key whose
+       hex representation exactly matches the given string, it and its value
+       are returned immediately, without whole database being searched."""
+    try:
+      p = bytes.fromhex(p_hex)
+      yield (p, self[p])
+    except Exception:
+      for k,v in self.items():
+        if k.hex().startswith(p_hex):
+          yield (k, v)
+  
+  def filterByMismatchedJSONObjValues(self, obj_ref, delete = False):
+    """This generator finds all entries ``k, v`` in the database for which
+       ``v``, when deserialized as JSON, does not match ``obj_ref`` - in the
+       sense that there is some key in both ``v`` and ``obj_ref`` which has
+       differing values in each (see ``mismatchedJSONObjValues``). Each such
+       ``k``, paired with ``v``'s list of mismatches (the list returned by
+       ``mismatchedJSONObjValues``), is then yielded. If ``delete`` is ``True``,
+       each entry is deleted from the database before being yielded."""
+    # NOTE: This function could just as well be defined as the body of
+    # ``onKVs`` below with ``self.items()`` for ``kvs`` and ``del self[k]``
+    # for `deleteFn`. However, that would result in creating an additional
+    # LMDB transaction for every deleted element. Instead, we have a special
+    # case for LMDB-backed databases which uses only a single tranaction.
+    def onKVs(kvs, deleteFn):
+      for k,v in kvs:
+        try:
+          ms = list(mismatchedJSONObjValues(json.loads(v), obj_ref))
+          if len(ms) > 0:
+            if delete: deleteFn(k)
+            yield (k, ms)
+        except Exception:
+          pass
+    if self._path is None:
+      yield from onKVs(self._impl.items(), self._impl.__delitem__); return
+    with self._impl.begin(write = True) as txn:
+      with txn.cursor() as curs:
+        yield from onKVs(curs, txn.delete)
+
+# ------------------------------------------------------------------------------
+# Helper functions for using LMDBOptDatabase objects with hex strings
+# ------------------------------------------------------------------------------
+
+def hex(hex_str):
+  """Convert a hex string to ``bytes``"""
+  return bytes.fromhex(hex_str)
+
+def jsonHex(obj_hex_str):
+  """Convert a hex string to ``bytes`` then deserialize a the result as JSON"""
+  return json.loads(hex(obj_hex_str))
+
+def pHex(v):
+  """Print the given value as a hex string, if it is a ``bytes`` object"""
+  if isinstance(v, bytes): print(v.hex())
+
+def pJSONHex(obj, pretty = False):
+  """Serialize the given object as JSON to get a ``bytes`` object then print
+     the result as a hex string"""
+  print(bytes(json.dumps(obj), 'utf-8').hex())
+
+def pHexPair(v1, v2, pretty = False):
+  """Print the given pair of values as a hex strings, if both are ``bytes``
+     objects. For the second element, if ``pretty`` is ``True``, deserialise it
+     as JSON and print that instead."""
```

Comment:
This `pretty` argument is rather interesting, as it completely changes the behavior of the function. I don't think this is necessarily a bad idea, but I wonder if it's worth having entirely separate functions for printing things as JSON instead of as hex. (Relatedly, I'm not sure where this `pretty` argument is actually used.)

Reply:
This and the other `pretty` were only there for manual debugging while I was writing this file. Removed since they're not used anymore.